# SQL FINAL PROJECT

## introduction
as a book app we want to get a few answers about the rating of books and authors for marketing the best salers so we will connect to the database and make a few quaeries to get our infomation.
first we will connect to the DB and learn our tables

In [1]:
#! pip install psycopg2

In [2]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

now with creating a loop we will explore the table that we can know them

In [3]:
table = ['books', 'reviews', 'ratings', 'publishers','authors']

for i in table:
    print("{} table".format(i))
    query = ''' SELECT *
                   FROM {}
                   LIMIT 10;
               '''.format (i)
    display(pd.read_sql_query(query, con=engine))

books table


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

reviews table


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

ratings table


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

publishers table


publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

authors table


author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

we explored our tables on the database, now let's start finding results for our tasks

## TASK 1
Find the number of books released after January 1, 2000.

In [4]:
query= ''' SELECT COUNT(title)
           FROM books
           WHERE publication_date > '2000-01-01';
        '''

In [5]:
t1= pd.io.sql.read_sql(query, con = engine)
t1

count
0    819

the answer for the first task is that we have 819 books that were released after 1/1/2000 

## TASK 2
Find the number of user reviews and the average rating for each book.

In [6]:
query=''' SELECT b.title,
                 COUNT(r.review_id) AS reviews,
                 AVG(ra.rating) AS rating
          FROM books AS b
          LEFT JOIN reviews AS r ON b.book_id = r.book_id
          LEFT JOIN ratings AS ra on b.book_id = ra.book_id
         GROUP BY b.book_id
         ORDER BY COUNT(r.review_id)DESC
          LIMIT 10 ;
        '''
          

In [7]:
t2=pd.io.sql.read_sql(query, con = engine)
t2

title  reviews    rating
0                            Twilight (Twilight  #1)     1120  3.662500
1                The Hobbit  or There and Back Again      528  4.125000
2                             The Catcher in the Rye      516  3.825581
3  Harry Potter and the Prisoner of Azkaban (Harr...      492  4.414634
4  Harry Potter and the Chamber of Secrets (Harry...      480  4.287500
5               Angels & Demons (Robert Langdon  #1)      420  3.678571
6  Harry Potter and the Order of the Phoenix (Har...      375  4.186667
7  The Lightning Thief (Percy Jackson and the Oly...      372  4.080645
8  The Fellowship of the Ring (The Lord of the Ri...      370  4.391892
9                                        Animal Farm      370  3.729730

in this table we can see the most reviews that books got and the top 5 are Twillight, The Hobbit, The catcher in the Rye and 2 Harry potter's. to check the rating is too hard in order with reviews so we will do that in the next table

In [8]:
query=''' SELECT b.title,
                 COUNT(r.review_id) AS reviews,
                 AVG(ra.rating) AS rating
          FROM books AS b
          LEFT JOIN reviews AS r ON b.book_id = r.book_id
          LEFT JOIN ratings AS ra on b.book_id = ra.book_id
         GROUP BY b.book_id
         ORDER BY AVG(ra.rating) DESC, COUNT(r.review_id) DESC
         LIMIT 10 ;
        '''
          

In [9]:
t2a=pd.io.sql.read_sql(query, con = engine)
t2a

title  reviews  rating
0                      A Dirty Job (Grim Reaper  #1)       16     5.0
1            School's Out—Forever (Maximum Ride  #2)       12     5.0
2       Moneyball: The Art of Winning an Unfair Game        9     5.0
3  In the Hand of the Goddess (Song of the Liones...        6     5.0
4                                              March        4     5.0
5                                    Light in August        4     5.0
6                                      Evening Class        4     5.0
7                                         Dead Souls        4     5.0
8  Wherever You Go  There You Are: Mindfulness Me...        4     5.0
9                    Act of Treason (Mitch Rapp  #9)        4     5.0

here we have the most rated books with a openninf high five with  A dirty job, school's out- forever, moneyball: the art of winning an unfair game, in the hand of a goddess and tai- pan.

## TASK 3
Identify the publisher that has released the greatest number of books with more than 50 pages 

In [10]:
query = ''' SELECT p.publisher,
                   COUNT(b.book_id)
            FROM books AS b
            LEFT JOIN publishers AS p ON b.publisher_id =p.publisher_id
            WHERE b.num_pages>50
            GROUP BY p.publisher
            ORDER BY COUNT(b.book_id) DESC
            LIMIT 8;
        '''
            

In [11]:
t3=pd.io.sql.read_sql(query, con = engine)
t3

publisher  count
0             Penguin Books     42
1                   Vintage     31
2  Grand Central Publishing     25
3          Penguin Classics     24
4          Ballantine Books     19
5                    Bantam     19
6                   Berkley     17
7        St. Martin's Press     14

and the most saller publushung for greater than 50 pages books is-- penguin books and right after them is vintage books.
this publishers are big and known so that maybe was obviyes

## TASK 4
Identify the author with the highest average book rating

In [12]:
query= ''' SELECT a.author,
                  AVG(r.rating)
           FROM books AS b
           LEFT JOIN authors AS a ON b.author_id=a.author_id
           LEFT JOIN ratings AS r ON b.book_id = r.book_id
           GROUP BY a.author
           HAVING COUNT(r.rating) >= 50
           ORDER BY AVG(r.rating) DESC;
           '''

In [13]:
t4= pd.io.sql.read_sql(query, con = engine)
t4

author       avg
0                                      Diana Gabaldon  4.300000
1                          J.K. Rowling/Mary GrandPré  4.288462
2                                     Agatha Christie  4.283019
3                   Markus Zusak/Cao Xuân Việt Khương  4.264151
4                                      J.R.R. Tolkien  4.240964
5                            Roald Dahl/Quentin Blake  4.209677
6                                   Louisa May Alcott  4.203704
7                                        Rick Riordan  4.130952
8                                       Arthur Golden  4.107143
9                                        Stephen King  4.009434
10                                       John Grisham  3.971429
11                                    William Golding  3.901408
12                                    Nicholas Sparks  3.882883
13                                       Jodi Picoult  3.881579
14                                    Sophie Kinsella  3.877193
15                                    James Patterson  3.859375
16                                      J.D. Salinger  3.846939
17           Paulo Coelho/Alan R. Clarke/Özdemir İnce  3.789474
18  William Shakespeare/Paul Werstine/Barbara A. M...  3.787879
19                                          Dan Brown  3.741259
20                                         Lois Lowry  3.738462
21            George Orwell/Boris Grabnar/Peter Škerl  3.729730
22                                    Stephenie Meyer  3.662500
23                                     John Steinbeck  3.643836

the most raited author is not less than  Diana Gabaldon (that i personlly never heard of) and after here comes the famoues J.K.Rowling and Mary Grandpre.

## TASK 5
 Find the average number of text reviews among users who rated more than 50 books.

In [14]:
query='''SELECT AVG(i.t) 
         FROM (SELECT COUNT(text) AS t, 
               username 
            FROM reviews GROUP BY username) AS i
         RIGHT JOIN (SELECT username 
                         FROM ratings 
                         GROUP BY ratings.username
                         HAVING COUNT(rating)>50) AS c ON i.username = c.username;
    '''

In [15]:
t5= pd.io.sql.read_sql(query, con = engine)
t5

avg
0  24.333333

the avrage text reviews for user how raites is 24.3 ratings 

so we looked up the DB and learned about books reviews and sales and know we have answers for marketing our app better